src/soc/ibm/power9/occ.c

```diff
@@ -0,0 +1,238 @@
+					break;
+
+				stream_push_control = read_scom(OCBSHCS_address);
+			}
+
+			if (counter == 4)
+				die("Failed to write to circular buffer.\n");
+		}
+	}
+}
+
+static void put_ocb_indirect(uint32_t ocb_req_length, uint32_t oci_address,
+			     uint64_t *ocb_buffer)
+{
+	write_scom(PU_OCB_PIB_OCBAR0, (uint64_t)oci_address << 32);
+
+	check_ocb_mode(PU_OCB_PIB_OCBCSR0_RO, PU_OCB_OCI_OCBSHCS0_SCOM);
+
+	for (uint32_t index = 0; index < ocb_req_length; index++)
+		write_scom(PU_OCB_PIB_OCBDR0, ocb_buffer[index]);
+}
+
+static void get_ocb_indirect(uint32_t ocb_req_length, uint32_t oci_address,
+			     uint64_t *ocb_buffer)
+{
+	write_scom(PU_OCB_PIB_OCBAR0, (uint64_t)oci_address << 32);
+	for (uint32_t loopCount = 0; loopCount < ocb_req_length; loopCount++)
+		ocb_buffer[loopCount] = read_scom(PU_OCB_PIB_OCBDR0);
+}
+
+void writeOCCSRAM(uint32_t address, uint64_t * buffer, size_t data_length)
+{
+	pm_ocb_setup(address);
+	put_ocb_indirect(data_length / 8, address, buffer);
+}
+
+void readOCCSRAM(uint32_t address, uint64_t * buffer, size_t data_length)
+{
+	pm_ocb_setup(address);
+	get_ocb_indirect(data_length / 8, address, buffer);
+}
+
+void write_occ_command(uint64_t write_data)
+{
+	check_ocb_mode(PU_OCB_PIB_OCBCSR1_RO, PU_OCB_OCI_OCBSHCS1_SCOM);
+	write_scom(PU_OCB_PIB_OCBDR1, write_data);
+}
+
+void clear_occ_special_wakeups(uint64_t cores)
+{
+	for (size_t i = 0; i < MAX_CORES_PER_CHIP; i += 2) {
+		if (!IS_EX_FUNCTIONAL(i, cores))
+			continue;
+		scom_and_for_chiplet(EC00_CHIPLET_ID + i, EX_PPM_SPWKUP_OCC, ~PPC_BIT(0));
+	}
+}
+
+void special_occ_wakeup_disable(uint64_t cores)
+{
+	enum { PPM_SPWKUP_FSP = 0x200F010B };
+
+	for (int i = 0; i < MAX_CORES_PER_CHIP; ++i) {
+		uint32_t spwkup_addr = PPM_SPWKUP_FSP + 0x01000000 * i;
+
+		if (!IS_EC_FUNCTIONAL(i, cores))
+			continue;
+
+		write_scom_for_chiplet(EC00_CHIPLET_ID + i, spwkup_addr, 0);
+		/* This puts an inherent delay in the propagation of the reset transition */
+		(void)read_scom_for_chiplet(EC00_CHIPLET_ID + i, spwkup_addr);
+	}
+}
+
+static uint32_t ppc_lis(uint16_t rt, uint16_t data)
+{
+	uint32_t inst;
+	inst = LIS_OP;
+	inst |= rt << (31 - 10);
+	inst |= data;
+	return inst;
+}
+
+static uint32_t ppc_ori(uint16_t rs, uint16_t ra, uint16_t data)
+{
+	uint32_t inst;
+	inst = ORI_OP;
+	inst |= rs << (31 - 10);
+	inst |= ra << (31 - 15);
+	inst |= data;
+	return inst;
+}
+
+static uint32_t ppc_mtspr(uint16_t rs, uint16_t spr)
+{
+	uint32_t temp = ((spr & 0x03FF) << (31 - 20));
+
+	uint32_t inst;
+	inst = MTSPR_OP;
+	inst |= rs << (31 - 10);
+	inst |= (temp & 0x0000F800) << 5;  // Perform swizzle
+	inst |= (temp & 0x001F0000) >> 5;  // Perform swizzle
+	return inst;
+}
+
+static uint32_t ppc_bctr(void)
+{
+	uint32_t inst;
+	inst = BCCTR_OP;
+	inst |= 20 << (31 - 10); // BO
+	return inst;
+}
+
+static uint32_t ppc_b(uint32_t target_addr)
+{
+	uint32_t inst;
+	inst = BR_OP;
+	inst |= (target_addr & 0x03FFFFFF);
+	return inst;
+}
```

Comment:
Please make those inline and move to ops.h. I can think of at least one use for ppc_mtspr elsewhere.